BUG: Fix groupby.apply() dropping _metadata from subclassed DataFrame

MengAiDev opened the pull request:
When extending pandas.DataFrame by subclassing, most operations preserve the _metadata attributes. This fix ensures that groupby.apply() also preserves these fields, making it consistent with other groupby operations like groupby.sum().

Fixes #62134

MengAiDev committed:
BUG: Fix groupby.apply() dropping _metadata from subclassed DataFrame

When extending pandas.DataFrame by subclassing, most operations preserve
the _metadata attributes. This fix ensures that groupby.apply() also
preserves these fields, making it consistent with other groupby operations
like groupby.sum().

Fixes #62134

rhshadrach reviewed (commented):
Thanks for the PR! Why not call `__finalize__` instead?

MengAiDev committed:
fix(groupby): preserve metadata for subclassed DataFrames and Series

- Update metadata preservation logic for DataFrames and Series in groupby operations
- Fix DataFrame.__setitem__ with MultiIndex columns and scalar indexer
- Adjust formatting and naming conventions in the code

MengAiDev committed:
fix(frame): ensure proper name attribute when setting new column with scalar value

- Address issue where name attribute was lost when setting new column with scalar value and tuple key in MultiIndex DataFrame
- Implement check for tuple key and scalar value in MultiIndex DataFrame
- Create Series with proper name to ensure name attribute matches the key
- Update test case to assert Series name in scalar assignment with MultiIndex

MengAiDev committed:
fix

rhshadrach commented:
Seems related to the change you're making here in `frame.py`.

rhshadrach commented:
Thanks for the PR, closing as stale for now. If you'd like to continue, fixup the failing tests and happy to reopen!
